### PR TITLE
Add `invoiceLineItem` and `batchInvoice` collections

### PR DESCRIPTION
**Pull request by dmccraw**

### CODE REVIEW

File: docs/collections/batch_invoices.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+* `params`: [int, object] - can be either a `teamId` or an object with query parameters.
+* `callback`: [function] - callback to be executed when the operation completes.
+
+To see a list of all available search params you can run:
+`teamsnap.collections.batchInvoices.queries.search.params`
+
+### Examples
+```javascript
+// ~~~~~
+// Loads all batchInvoices for `teamId: 1`.
```
Comment (matthopson) [inline docs/collections/batch_invoices.md]: Should this be `division`?

File: docs/collections/batch_invoices.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+teamsnap.loadBatchInvoices(1);
+
+// ~~~~~
+// Loads all batchInvoices for `batchInvoiceId: 1`.
+teamsnap.loadBatchInvoices({batchInvoiceId: 1});
```
Comment (matthopson) [inline docs/collections/batch_invoices.md]: Should this be `id: 1` ?

File: docs/collections/batch_invoices.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+- [deleteBatchInvoice](#deleteBatchInvoice)
+
+
+---
+<a id="loadBatchInvoices"></a>
+## `loadBatchInvoices(params, callback)`
+Loads items from the `batchInvoices` collection based on given params.
+
+### Params
+* `params`: [int, object] - can be either a `teamId` or an object with query parameters.
```
Comment (matthopson) [inline docs/collections/batch_invoices.md]: `divisionId`

File: docs/collections/batch_invoices.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+`teamsnap.collections.batchInvoices.queries.search.params`
+
+### Examples
+```javascript
+// ~~~~~
+// Loads all batchInvoices for `teamId: 1`.
+teamsnap.loadBatchInvoices(1);
+
+// ~~~~~
+// Loads all batchInvoices for `batchInvoiceId: 1`.
```
Comment (matthopson) [inline docs/collections/batch_invoices.md]: `id`

File: docs/collections/batch_invoices.md (hunk 3 above)
Comment (matthopson) [inline docs/collections/batch_invoices.md]: Also, is this functionality never going to be available to teams? Generally we have a `league_` designation for collections like these.

File: docs/collections/invoice_line_items.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+- [deleteInvoiceLineItem](#deleteInvoiceLineItem)
+
+
+---
+<a id="loadInvoiceLineItems"></a>
+## `loadInvoiceLineItems(params, callback)`
+Loads items from the `invoiceLineItems` collection based on given params.
+
+### Params
+* `params`: [int, object] - can be either a `batchInvoiceId` or an object with query parameters.
```
Comment (matthopson) [inline docs/collections/invoice_line_items.md]: We never assume the `id` being passed is related to an object other than a `team` or `division`. Should be `teamId` or `divisionId`

File: docs/collections/invoice_line_items.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+* `callback`: [function] - callback to be executed when the operation completes.
+
+To see a list of all available search params you can run:
+`teamsnap.collections.invoiceLineItems.queries.search.params`
+
+### Examples
+```javascript
+// ~~~~~
+// Loads all invoiceLineItems for `batchInvoiceId: 1`.
```
Comment (matthopson) [inline docs/collections/invoice_line_items.md]: Should be `id: 1`

File: docs/collections/invoice_line_items.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+teamsnap.loadInvoiceLineItems(1);
+
+// ~~~~~
+// Loads all invoiceLineItems for `invoiceLineItemId: 1`.
+teamsnap.loadInvoiceLineItems({invoiceLineItemId: 1});
```
Comment (matthopson) [inline docs/collections/invoice_line_items.md]: Should be `id: 1`

File: docs/collections/invoice_line_items.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+// Creates a new invoiceLineItem, saves, then deletes it.
+var invoiceLineItem = teamsnap.createInvoiceLineItem({
+  batchInvoiceId: 1,
+  quantity: '1',
+  amount: 1.00
+});
+
+teamsnap.saveInvoiceLineItem(invoiceLineItem).then(function(){
+  // Save complete, now delete.
+  teamsnap.deleteInvoiceLineItem(invoiceLineItem).then(function(){
```
Comment (matthopson) [inline docs/collections/invoice_line_items.md]: You should be deleting what's being returned in the save action:
```
teamsnap.saveInvoiceLineItem(invoiceLineItem).then(function(savedItem){
teamsnap.deleteInvoiceLineItem(savedItem).then(function(){.....
```

File: src/collections/invoiceLineItems.coffee [hunk 9]
```diff
@@ -0,0 +1,41 @@
+# Load invoiceLineItems by batchInvoiceId or query parameters
+exports.loadInvoiceLineItems = (params, callback) ->
+  if @isId params
+    params = batchInvoiceId: params
```
Comment (matthopson) [inline src/collections/invoiceLineItems.coffee]: See note in docs - we should be using a `teamId` or `divisionId` as the default parameter.

File: src/collections/invoiceLineItems.coffee [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+exports.saveInvoiceLineItem = (invoiceLineItem, callback) ->
+  unless invoiceLineItem
+    throw new TSArgsError 'teamsnap.saveInvoiceLineItem',
+      "`invoiceLineItem` must be provided"
+  unless @isItem invoiceLineItem, 'invoiceLineItem'
+    throw new TSArgsError 'teamsnap.saveInvoiceLineItem',
+      "`invoiceLineItem.type` must be 'invoiceLineItem'"
+  unless invoiceLineItem.batchInvoiceId
+    return @reject 'You must choose a division.', 'batchInvoiceId', callback
```
Comment (matthopson) [inline src/collections/invoiceLineItems.coffee]: Should say `You must choose a batchInvoiceId`

**Comment by matthopson** [comment]
Other things:
- Finish tests
- Add to `types`

File: docs/collections/batch_invoices.md (hunk 3 above)
Comment (dmccraw) [inline docs/collections/batch_invoices.md]: BatchInvoice will probably be available on the team side in the future.

File: docs/collections/invoice_line_items.md (hunk 5 above)
Comment (dmccraw) [inline docs/collections/invoice_line_items.md]: LineItems belong to a BatchInvoice though :/ They don't live on a team or a division. Thoughts?

File: src/collections/invoiceLineItems.coffee (hunk 9 above)
Comment (dmccraw) [inline src/collections/invoiceLineItems.coffee]: The problem is that Line Items are associated with a Batch Invoice or a Invoice not a team or a  division.

File: docs/collections/invoice_line_items.md (hunk 5 above)
Comment (matthopson) [inline docs/collections/invoice_line_items.md]: Every item we handle is either scoped to a team or division. I would expect these to be as well. How would you bulk_load them? Think about `statistic_data` - they technically belong to an `event_statistic`, but we still scope them to team.

File: docs/collections/batch_invoices.md (hunk 3 above)
Comment (matthopson) [inline docs/collections/batch_invoices.md]: Then we probably need to re-think a lot of these references. Most of them point to `divisionId` as a default param.

File: src/collections/invoiceLineItems.coffee (hunk 9 above)
Comment (matthopson) [inline src/collections/invoiceLineItems.coffee]: But shouldn't they be? All items are (except for maybe `sports`, `time_zones`, etc...)

File: docs/collections/batch_invoices.md (hunk 3 above)
Comment (matthopson) [inline docs/collections/batch_invoices.md]: I feel like this should have be:
`batch_invoices`, `league_batch_invoices`, etc... to account for teams and leagues... and they could accept a team or division id as the default param (respectively).

Comment (dmccraw) [inline docs/collections/batch_invoices.md]: They eventually will be on both the division and the team. What about not having a default param but making them specify?

File: docs/collections/invoice_line_items.md (hunk 5 above)
Comment (dmccraw) [inline docs/collections/invoice_line_items.md]: I am not sure we we want to bulk load them. Long term a line item will be available through a batch invoice, invoice, division or team (division/team maybe). Not sure the best way to handle such a flexible item.

File: docs/collections/batch_invoices.md (hunk 1 above)
Comment (dmccraw) [inline docs/collections/batch_invoices.md]: fixed

File: docs/collections/batch_invoices.md (hunk 2 above)
Comment (dmccraw) [inline docs/collections/batch_invoices.md]: fixed

File: docs/collections/batch_invoices.md (hunk 4 above)
Comment (dmccraw) [inline docs/collections/batch_invoices.md]: fixed

File: src/collections/invoiceLineItems.coffee (hunk 9 above)
Comment (dmccraw) [inline src/collections/invoiceLineItems.coffee]: I will go ahead and add divisionId to lineItem. We can in apiv3 load it through all the batchInvoices for that division.

File: src/collections/invoiceLineItems.coffee (hunk 10 above)
Comment (dmccraw) [inline src/collections/invoiceLineItems.coffee]: fixed

File: docs/collections/invoice_line_items.md (hunk 6 above)
Comment (dmccraw) [inline docs/collections/invoice_line_items.md]: changed to `divisionId`

**Comment by dmccraw** [comment]
@matthopson Everything should be fixed except for the tests.

**Comment by dmccraw** [comment]
ok, removed the tests :( and changed the version to 1.29.0

**Comment by dmccraw** [comment]
@matthopson all updated

**Comment by dmccraw** [comment]
@matthopson Updated, remove the default param of divisionId so requires an object.

**Comment by dmccraw** [comment]
@matthopson Can you release this to npm for me?

**Pull request merged**